apps/src/javalab/JavalabFileManagement.js

```diff
@@ -27,24 +25,29 @@ function renameProjectFile(filename, newFilename) {
   );
 }
 
-function onSave() {
-  header.showProjectSaving();
-  const filename = getStore().getState().javalab.filename;
-  const editorText = getStore().getState().javalab.editorText;
+function onSave(success, failure) {
   // TODO: enable multi-file
-  filesApi.putFile(
-    filename,
-    editorText,
-    () => {
-      // at this point project.filesVersionId will be updated and we can save
-      // sources as well.
-      project.save();
-    },
-    () => {
-      console.warn(`Javalab: error file ${filename} not saved`);
-      project.showSaveError_();
+  const javalabState = getStore().getState().javalab;
+  // only save file if file content has changed.
+  if (javalabState.filesChanged) {
```

Comment:
this should (and lines 34 & 35 below where we access redux state) probably be a selector within javalabRedux. there are 2 ways i've seen this done in our codebase:

```javascript
// option 1
export getFilesChanged() {
  return getStore().getState().javalab.filesChanged;
}

// option 2
export getFilesChanged(state) {
  return state.javalab.filesChanged;
}
```

option 2 is nice because it's easier to unit test, but otherwise either option is fine